segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,308 @@
+        return getLastProcessedEventOffset(eventProcessor.segment)
+                .thenApplyAsync(currentOffset -> readEventsFromSegment(eventProcessor, currentOffset), this.executor);
+    }
+
+    private EventsReadAndTruncationPoints readEventsFromSegment(EventProcessorImpl eventProcessor, int currentOffset) {
+        List<BufferView> readEvents = new ArrayList<>();
+        int finalTruncationOffset = currentOffset;
+        while (readEvents.size() < eventProcessor.getConfig().getMaxItemsAtOnce()) {
+            try {
+                int appendLength = bufferViewToInt(readAppendOfLength(eventProcessor.segment, finalTruncationOffset, Integer.BYTES));
+                finalTruncationOffset += Integer.BYTES;
+                BufferView data = readAppendOfLength(eventProcessor.segment, finalTruncationOffset, appendLength);
+                readEvents.add(data);
+                finalTruncationOffset += data.getLength();
+            } catch (Exception e) {
+                // We got some data, do not throw, just stop looping.
+                if (!readEvents.isEmpty()) {
+                    break;
+                }
+                // TimeoutException is expected if there is no data in the Segment. Otherwise it may be a problem.
+                throw (e instanceof TimeoutException) ? new NoDataAvailableException() : new CompletionException(e);
+            }
+        }
+        return new EventsReadAndTruncationPoints(readEvents, currentOffset, finalTruncationOffset);
+    }
+
+    private CompletableFuture<Void> updateAndTruncateInternalSegment(EventProcessorImpl ep,
+                                                                     EventsReadAndTruncationPoints readResult,
+                                                                     Timer iterationTime) {
+        return updateLastProcessedEventOffset(ep.segment, readResult.getFinalTruncationOffset())
+                .thenCompose(v -> ep.segment.truncate(readResult.getFinalTruncationOffset(), OPERATION_TIMEOUT))
+                .thenAccept(v -> {
+                    // Decrement outstanding bytes and publish the last latency value after successful truncation.
+                    ep.outstandingBytes.addAndGet(-readResult.getReadBytes());
+                    ep.lastIterationLatency.set(iterationTime.getElapsedMillis());
+                });
+    }
+
+    private BufferView readAppendOfLength(DirectSegmentAccess segment, int offset, int length) throws Exception {
+        @Cleanup
+        ReadResult readResult = segment.read(offset, length, OPERATION_TIMEOUT);
+        ReadResultEntry entry = readResult.next();
+        entry.requestContent(OPERATION_TIMEOUT);
+        return entry.getContent().get(OPERATION_TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    private int bufferViewToInt(BufferView appendLengthContent) throws IOException {
+        return ByteBuffer.wrap(appendLengthContent.getReader().readAllBytes()).getInt();
```

Comment:
`BufferView.getBufferViewReader().readInt()` - avoids copying into new array.

Reply:
This is not necessary now, removed.